// @ngageoint/geopackage/built/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var proj4Defs_1 = __importDefault(require("./lib/proj4Defs"));
exports.proj4Defs = proj4Defs_1.default;
var crsWkt_1 = require("./lib/extension/crsWkt");
exports.CrsWktExtension = crsWkt_1.CrsWktExtension;
var geoPackageConnection_1 = require("./lib/db/geoPackageConnection");
exports.GeoPackageConnection = geoPackageConnection_1.GeoPackageConnection;
var webp_1 = require("./lib/extension/webp");
exports.WebPExtension = webp_1.WebPExtension;
var rtreeIndex_1 = require("./lib/extension/rtree/rtreeIndex");
exports.RTreeIndex = rtreeIndex_1.RTreeIndex;
var metadata_1 = require("./lib/extension/metadata");
exports.MetadataExtension = metadata_1.MetadataExtension;
var dataColumnsDao_1 = require("./lib/dataColumns/dataColumnsDao");
exports.DataColumnsDao = dataColumnsDao_1.DataColumnsDao;
var mediaTable_1 = require("./lib/extension/relatedTables/mediaTable");
exports.MediaTable = mediaTable_1.MediaTable;
var userMappingTable_1 = require("./lib/extension/relatedTables/userMappingTable");
exports.UserMappingTable = userMappingTable_1.UserMappingTable;
var dataColumnConstraintsDao_1 = require("./lib/dataColumnConstraints/dataColumnConstraintsDao");
exports.DataColumnConstraintsDao = dataColumnConstraintsDao_1.DataColumnConstraintsDao;
var featureColumn_1 = require("./lib/features/user/featureColumn");
exports.FeatureColumn = featureColumn_1.FeatureColumn;
var userColumn_1 = require("./lib/user/userColumn");
exports.UserColumn = userColumn_1.UserColumn;
var tileColumn_1 = require("./lib/tiles/user/tileColumn");
exports.TileColumn = tileColumn_1.TileColumn;
var dataColumns_1 = require("./lib/dataColumns/dataColumns");
exports.DataColumns = dataColumns_1.DataColumns;
var dataTypes_1 = require("./lib/db/dataTypes");
exports.DataTypes = dataTypes_1.DataTypes;
var schema_1 = require("./lib/extension/schema");
exports.SchemaExtension = schema_1.SchemaExtension;
var geometryColumns_1 = require("./lib/features/columns/geometryColumns");
exports.GeometryColumns = geometryColumns_1.GeometryColumns;
var metadataReference_1 = require("./lib/metadata/reference/metadataReference");
exports.MetadataReference = metadataReference_1.MetadataReference;
var geometryData_1 = require("./lib/geom/geometryData");
exports.GeometryData = geometryData_1.GeometryData;
var tableCreator_1 = require("./lib/db/tableCreator");
exports.TableCreator = tableCreator_1.TableCreator;
var dublinCoreType_1 = require("./lib/extension/relatedTables/dublinCoreType");
exports.DublinCoreType = dublinCoreType_1.DublinCoreType;
var boundingBox_1 = require("./lib/boundingBox");
exports.BoundingBox = boundingBox_1.BoundingBox;
var api_1 = require("./lib/api");
exports.GeoPackage = api_1.GeoPackageAPI;
var metadata_2 = require("./lib/metadata/metadata");
exports.Metadata = metadata_2.Metadata;
var features_1 = require("./lib/tiles/features");
exports.FeatureTiles = features_1.FeatureTiles;
var numberFeaturesTile_1 = require("./lib/tiles/features/custom/numberFeaturesTile");
exports.NumberFeaturesTile = numberFeaturesTile_1.NumberFeaturesTile;
var shadedFeaturesTile_1 = require("./lib/tiles/features/custom/shadedFeaturesTile");
exports.ShadedFeaturesTile = shadedFeaturesTile_1.ShadedFeaturesTile;
var retriever_1 = require("./lib/tiles/retriever");
exports.GeoPackageTileRetriever = retriever_1.GeoPackageTileRetriever;
var tileUtilities_1 = __importDefault(require("./lib/tiles/creator/tileUtilities"));
exports.TileUtilities = tileUtilities_1.default;
