// @ngageoint/geopackage/built/lib/db/dataTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DataTypes;
(function (DataTypes) {
    DataTypes[DataTypes["BOOLEAN"] = 0] = "BOOLEAN";
    DataTypes[DataTypes["TINYINT"] = 1] = "TINYINT";
    DataTypes[DataTypes["SMALLINT"] = 2] = "SMALLINT";
    DataTypes[DataTypes["MEDIUMINT"] = 3] = "MEDIUMINT";
    DataTypes[DataTypes["INT"] = 4] = "INT";
    DataTypes[DataTypes["INTEGER"] = 5] = "INTEGER";
    DataTypes[DataTypes["FLOAT"] = 6] = "FLOAT";
    DataTypes[DataTypes["DOUBLE"] = 7] = "DOUBLE";
    DataTypes[DataTypes["REAL"] = 8] = "REAL";
    DataTypes[DataTypes["TEXT"] = 9] = "TEXT";
    DataTypes[DataTypes["BLOB"] = 10] = "BLOB";
    DataTypes[DataTypes["DATE"] = 11] = "DATE";
    DataTypes[DataTypes["DATETIME"] = 12] = "DATETIME";
    DataTypes[DataTypes["GEOMETRY"] = 13] = "GEOMETRY";
})(DataTypes = exports.DataTypes || (exports.DataTypes = {}));
// eslint-disable-next-line @typescript-eslint/no-namespace
(function (DataTypes) {
    function nameFromType(type) {
        return DataTypes[type];
    }
    DataTypes.nameFromType = nameFromType;
    function fromName(type) {
        return DataTypes[type];
    }
    DataTypes.fromName = fromName;
})(DataTypes = exports.DataTypes || (exports.DataTypes = {}));
